### PR TITLE
perf(hooks): cmdline-first Linux db-lock scan, drop the lsof fallback (#2180)

### LINKED ISSUE #2180: perf(hooks): replace the O(procs×fds) Linux db-lock scan with a cmdline-first sweep and drop the lsof fallback (proposal)
## Area

Claude Code hooks (`gitnexus/hooks/claude/hook-db-lock-probe.cjs`) — Linux db-lock probe

## Context

#2163 → #2165 → #2169 contained the orphan storm: probe and augment subprocesses are now guarded, concurrency is slot-bounded, and a busy host can no longer snowball to 100% CPU. What those PRs deliberately did **not** change is the probe's detection algorithm — and on busy hosts it still cannot answer the question it exists for:

- `linuxProcScanFindGitNexusServer` walks `/proc/<pid>/fd/*` with `stat(2)` for **every process** — O(processes × fds). Measured on the incident-class host (~660 processes / ~18,300 fds): **~280 ms** for a full scan on an idle box, and a guaranteed `GITNEXUS_HOOK_LINUX_PROC_BUDGET_MS` (1200 ms) blowout under load.
- The `lsof` fallback then burns its full 1 s `spawnSync` budget (real `lsof` takes ~2 s there) and lands ETIMEDOUT → fail-closed.

Net effect on busy hosts: every Grep/Glob/Bash hook invocation spends ~2.2 s of CPU to conclude "couldn't tell" and skip the augment. The guards from #2165/#2169 make this safe; this proposal makes it *fast and correct*.

## Proposal: invert the scan — identify candidates by name first, stat only their fds

Rewrite `linuxProcScanFindGitNexusServer` (name kept) as a tri-state (`'owned' | 'not-owned' | 'timeout'`) three-phase scan:

1. **Phase 0 — `/proc/<pid>/comm`**: one tiny read per process (kernel serves it from `task->comm`, never touches the target's memory maps; microseconds each). Keep only processes whose comm could plausibly be a Node/gitnexus runner (`node`, `gitnexus`, `bun`, `npm`, …).
2. **Phase 1 — `/proc/<pid>/cmdline`** for survivors only, read via `openSync`+`readSync` capped at 4 KB (guards against D-state processes wedged on `mmap_lock` and against giant cmdlines). Prefilter with the existing `isGitNexusServerCommand`.
3. **Phase 2 — fd stat** for the 0–2 remaining candidates only: compare `dev`+`ino` against the lbug file, exactly as today.

Measured prototype on the same host: **34 ms (phase 0+1 full sweep) + 2 ms (candidate fd stat)** vs 280 ms today — comfortably inside the budget even under heavy load, with answers identical to `lsof`'s.

**Delete the Linux `lsof` fallback entirely.** `lsof` is itself just a procfs reader — it has no incremental detection power over the scan above, and after #2165 it only ever times out on the hosts where it would matter. (macOS keeps its `lsof`+`ps` path; Windows keeps Restart Manager. Linux-only change.)

## The one contract change (why this needs eyes before code)

Budget expiry currently returns `false` from the proc scan and falls through to `lsof`; post-change, `'timeout'` maps to **fail-closed** (treat as owned → skip augment), and the header fail-matrix gets rewritten accordingly. Two notes:

- On busy hosts the *end-to-end* semantics today are already fail-closed (via the lsof ETIMEDOUT arm) — the change removes the 1–2 s of wasted work, not the outcome. The hosts that genuinely change behavior are mid-load Linux boxes where the old fd-walk expired but `lsof` would have answered within 1 s: those now self-throttle (skip augment) instead. Overload self-throttling is precisely what the #2163 incident needed.
- While in there: fix the budget parsing bug — `Number(raw && String(raw).trim())` makes an explicit `"0"` fall back to 1200; proposed semantics `<= 0` → immediate `'timeout'`, which doubles as a deterministic test vector.

## Test plan / merge conditions

- New env `GITNEXUS_HOOK_PROC_ROOT` as a fake-procfs injection point; the test harness defaults to a fixture root so Linux unit tests stop sweeping the real `/proc`. Existing owner tests keep their assertions, only the injection changes; new `hook-db-lock-probe.test.ts` covers the edge cases and registers in the cross-platform suite.
- **Live e2e as a merge condition**: spawn a real `gitnexus mcp` (or a minimal real-lbug opener), wait for it to hold the lbug, assert the proc scan reports `owned`. Rationale: "the lbug handle is fd-visible" is an implementation property of `@ladybugdb/core` with zero CI protection today — a future switch to mmap-only I/O would silently blind the probe and regress #1492 with no red test. Cheap belt: also match candidates' `/proc/<pid>/maps` against the db path.
- Edge rulings (each pinned by a test): candidate's `/proc/<pid>/fd` unreadable (EACCES, cross-user server) → fail-closed `'owned'`, documented tradeoff; pathological hosts (tens of thousands of processes, ~1.5 s linear extrapolation) → `'timeout'` → fail-closed, documented boundary.

## Out of scope / related follow-ups (separate items)

- TTL cache for probe results (the other half of the perf story; tri-state-aware — `timeout` results must be cacheable since that's the load-shedding path).
- Cursor integration probe (it installs no probe sibling today; #2169 left its augment unwrapped for the same reason).
- Guard self-test does not verify group-kill capability (busybox `timeout` passes the exit-propagation test but cannot reap an npx grandchild — noted in #2169's review).
- `isGitNexusServerCommand` dev-checkout blind spot (`node …/gitnexus/dist/cli/index.js mcp` doesn't match; pre-existing, affects old and new scan equally).

Happy to implement this (same flow as #2165/#2169) once the direction — particularly the fail-closed-on-timeout mapping and the lsof-fallback removal — has maintainer sign-off.

### PR DESCRIPTION
**Pull request by Minidoracat**
## Summary

Replaces the O(processes × fds) Linux db-lock scan with a comm→cmdline→fd three-phase sweep and removes the Linux `lsof` fallback, as proposed in #2180. On a busy host the old scan blew its budget and fell through to `lsof` (~2 s → ETIMEDOUT → fail-closed); every Grep/Glob/Bash hook spent ~2 s of CPU to conclude "couldn't tell." macOS (`lsof`+`ps`) and Windows (Restart Manager) paths are untouched.

Closes #2180. Final piece of the #2163 follow-up list (cmdline-first scan); the probe-result TTL cache and cursor-probe items remain separate future work.

Measured on a 583-process host with 6 background `gitnexus mcp` servers: **owner detection 6–12 ms, was ~1216 ms + lsof timeout (~100×)**.

## Areas touched

- [x] `gitnexus/` (CLI / core / MCP server) — hooks + tests only
- [ ] others

Also `gitnexus-claude-plugin/hooks/` (byte-identical probe copy).

## What changed

- `linuxProcScanFindGitNexusServer` (name kept — source-pinned; return type now tri-state `'owned' | 'not-owned' | 'timeout'`):
  - **Phase 0** — `/proc/<pid>/comm` prefilter. Kernel `task->comm`, never touches the target's memory maps, so it can't wedge on a D-state process's `mmap_lock`. Truncation-safe whitelist match (comm is capped at 15 visible chars).
  - **Phase 1** — bounded `/proc/<pid>/cmdline` read (`openSync`+`readSync`, default 16 KiB, floor 4 KiB, bounded escalation to a hard ceiling) for survivors only, prefiltered by `isGitNexusServerCommand`.
  - **Phase 2** — `dev`+`ino` fd match for the 0–2 remaining candidates only.
- Dispatch: `'owned'` and `'timeout'` → `true`; `'not-owned'` → `false`. **The Linux `lsof` fallback is removed.**
- Budget parse fix: the old `Number(raw && trim()) ? … : 1200` treated `"0"` as falsy and fell back to 1200; now parse-then-validate, with `<= 0` an explicit immediate timeout (also a deterministic test vector).
- `GITNEXUS_HOOK_PROC_ROOT` (default `/proc`, unset ⇒ zero behavior change) so the Linux scan is unit-testable against a fixture procfs instead of the host's real `/proc`.

## Two tradeoffs worth your eyes

**1. `timeout` → fail-closed (the one contract change).** Budget expiry now maps to "treat as owned, skip augment" instead of falling through to `lsof`. End-to-end on busy hosts this is unchanged — the old `lsof` arm also fail-closed there (ETIMEDOUT), just 2 s slower and spawning the orphan this whole #2163 line was about. The hosts that *change* are mid-load Linux boxes where the old fd-walk expired but `lsof` would have answered within 1 s: those now self-throttle (skip that augment) rather than spend the time. With the 1200 ms default and 6–12 ms typical scans there's ~100× headroom, so only genuinely overloaded (incident-class) hosts reach it — which is exactly where self-throttling is wanted.

**2. Phase 0 comm whitelist is calibrated to `@ladybugdb/core`.** Empirically, a real `gitnexus mcp` server reports `comm = MainThread` (its `worker_threads` setup renames the main thread), **not** `node` — so `MainThread` is whitelisted alongside the launcher basenames; omitting it would blind the probe to every server (#1492-class owner miss). The match is bidirectional-prefix (truncation-safe) and Phase 0 is deliberately over-collecting (a stray candidate only costs one bounded cmdline read; Phase 2's `dev`+`ino` is the authority). The residual: if a future `@ladybugdb`/Node build renames that thread to something whose 15-char prefix matches no whitelist entry, Phase 0 would filter real servers out cheaply and the live e2e (which runs the current build) wouldn't catch it. Documented at the whitelist with a maintenance note; flagging here because it couples this probe to a dependency's thread naming. If you'd prefer a comm-independent prefilter (e.g. read cmdline for all procs — the measured full sweep is still ~34 ms — at the cost of touching unrelated procs' mm), that's a design call I'm happy to take either way.

## Testing & verification

- [x] `cd gitnexus && npx vitest run test/unit/hook-db-lock-probe.test.ts test/unit/hooks.test.ts` — 165 passed, 0 failed (the relaned lsof suites skip on Linux)
- [x] new `hook-db-lock-probe.test.ts` drives all three phases against a fake procfs: comm-truncation safety, Phase 0 trap (cmdline looks like a server but comm doesn't → filtered), 4 KiB-boundary owner-miss guard, `budget=0` immediate timeout, EACCES → fail-closed
- [x] **live-`/proc` e2e** pins the "lbug handle is fd-visible" property of `@ladybugdb` against a real subprocess holder, asserting only on the holder we spawned (background servers on the host can't false-positive a fresh temp lbug)
- [x] `resolve-invocation` / `setup` / `cli-commands` / `antigravity-hook-e2e` — green
- [x] probe copies `diff -q` byte-identical; `windowsHide` parity holds (no new spawns)
- [x] budget-parse red/green: base treats `"0"` as 1200; fixed → immediate timeout

**D5 — relaned lsof tests (the delicate bit).** On this incident-class host the two pre-existing `#1493` "ENOENT lsof → fail-open" tests were already red on base (real `lsof` exceeds its 1 s budget → ETIMEDOUT → fail-closed). Since Linux no longer takes the `lsof` path, the `lsof`/`ps` owner-detection suites are relaned to macOS (assertions unchanged, still guarding the path where it's live), and the `lsof` orphan-reaping suite is removed with a rationale note (no `lsof` is spawned on Linux now; wrapper-reaping stays covered by the augment-reaping suites and the macOS lsof lane). All Linux owner-detection correctness moved to the fake-procfs tests above.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: gitnexus/hooks/claude/hook-db-lock-probe.cjs [hunk 1]
```diff
@@ -242,59 +264,226 @@ function hasGitNexusServerOwnerWindows(dbPathAbs, myPid) {
...
-      continue;
+    } catch (err) {
+      // A candidate already matched the GitNexus-server cmdline, so an
+      // unreadable fd dir is almost always a cross-user server (EACCES): we
+      // cannot prove it does NOT hold the lbug, and the candidate is a strong
+      // owner signal. Fail-closed (treat as owner) for any error other than
+      // ENOENT (the process raced away -> genuinely not an owner). Documented
+      // tradeoff (#2180 edge): a cross-user gitnexus server skips augment.
+      if (err && err.code === 'ENOENT') continue;
+      return 'owned';
```
Comment (magyargergo) [inline gitnexus/hooks/claude/hook-db-lock-probe.cjs]: **[P2 · code-read · Codex + 4 consistent Claude lanes] Phase-2 fd-dir `EACCES → 'owned'` over-skips on multi-user / root hosts.**

`/proc/<pid>/comm` + `/cmdline` are world-readable (0444) but `/proc/<pid>/fd` is owner-only (0500, ptrace-gated). A **cross-user or root-run `gitnexus mcp`/`serve` process serving *any* repo** passes Phase 0+1, then `readdirSync(fdDir)` throws `EACCES` here → this `return 'owned'` fires → the dispatcher (`verdict !== 'not-owned'`) returns `true` → this repo's augment is silently skipped, even though that server never holds this repo's lbug. The old code did `catch { continue; }` and ran the cmdline check *after* the fd-holds check, so it did not over-skip this way.

It's fail-closed (safe direction) and the comment acknowledges the cross-user case, but: (a) the candidate was never shown to hold *this* lbug (dev+inode check skipped on this path); (b) the catch returns `'owned'` for **all** non-`ENOENT` errors (`ENOTDIR`/`EIO`/`EPERM`/`ESTALE`), some transient; (c) no diagnostic — the adapter's `GITNEXUS_DEBUG` line says "MCP server owns DB" for owned/timeout/this case alike.

Suggested: `return 'timeout'` on `EACCES` (same fail-closed outcome, no false ownership claim) and/or narrow to permission errors only + a distinct `GITNEXUS_DEBUG` reason. Mirror in the byte-identical `gitnexus-claude-plugin/hooks/hook-db-lock-probe.cjs`.

File: gitnexus/test/unit/hook-db-lock-probe.test.ts [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+        { pid: 9201, comm: 'bash', cmdline: ['bash', '-l'], fdTargets: [] },
+      ],
+      { GITNEXUS_HOOK_LINUX_PROC_BUDGET_MS: '5000' },
+    );
+    expect(owned).toBe(false);
+  });
+
+  // ── D3: 4 KB+ cmdline cap ─────────────────────────────────────────
+
+  it('owned even when the mode token sits far past 4 KB (read escalation, no miss)', () => {
```
Comment (magyargergo) [inline gitnexus/test/unit/hook-db-lock-probe.test.ts]: **[P2 · code-read · test integrity · 3 consistent Claude lanes] This test does not exercise the escalation path it claims to cover.**

`longPad = '/very/long/path'.repeat(600)` is ~9 KB; the test sets no `GITNEXUS_HOOK_PROC_CMDLINE_MAX`, so the cap is the 16 KB default (`getCmdlineMaxBytes`). 9 KB < 16 KB → the first `readSync` reads the whole cmdline in one call and `readLinuxCmdline`'s multi-iteration escalation loop never repeats. So **no test in the suite actually triggers escalation** — the verdict here is reached via the single-read path.

The escalation code itself is correct (independently verified), so this is a coverage/labeling gap rather than a live bug, but the path that exists specifically to avoid missing a far-out `mcp`/`serve` token is unguarded against regression. Fix in-PR: add `GITNEXUS_HOOK_PROC_CMDLINE_MAX: '4096'` to the env (with the `mcp` token past the 4 KB cap) so the loop actually iterates — or rename the test to reflect what it tests.

**Review by magyargergo — COMMENTED** [review]
## Tri-review: PR #2183 — cmdline-first Linux db-lock scan (#2180)

**Methods (3):** GitNexus swarm (risk, test/CI) · Compound-Engineering personas (correctness, adversarial, performance, reliability) · Codex. **Engine breakdown:** 6 Claude-persona lanes + **Codex, the one genuinely independent engine** (live this run after a foreground retry — the first background dispatch evaporated). Claude-lane agreement below is *consistent corroboration*, not independent confirmation; only a **Codex-and-Claude** agreement is cross-engine. Reviewed through the requested **Linux OS-semantics + performance** lens. Both `hook-db-lock-probe.cjs` copies (claude + plugin) are byte-identical.

### Validated / solid (credit where due)
- **Performance headline holds** — Codex *and* the performance lane independently confirm a genuine **~40× syscall reduction** (old `O(procs×fds)` stat-storm + lsof spawn → `O(procs)` cheap `comm` reads + `O(few)` bounded `cmdline` reads + `O(0–2)` fd stats), with **no input slower than before**. The CHANGELOG "~1.2 s+ETIMEDOUT → ~12 ms" claim is substantiated.
- **Budget-parse fix is correct** — the old `Number("0")`-is-falsy → 1200 ms bug is genuinely fixed; the tri-state dispatcher `verdict !== 'not-owned'` is exhaustive (both `owned` and `timeout` fail closed).
- **No owner-miss** via the Phase-0 `comm` whitelist for any production launch shape (`node`/`gitnexus`/`MainThread`, 15-char-truncation-safe) — refuted by correctness, adversarial, *and* Codex.
- Escalation regexes, short-read handling, TOCTOU, ReDoS all probed and **refuted**; the deleted Linux lsof-orphan tests are correctly removed (no lsof leg remains on Linux); the CHANGELOG edit is the root (non-release-owned) file — `gitnexus/CHANGELOG.md` is untouched — **no policy violation**.

### ① Headline (inline, P2 — Codex + 4 consistent Claude lanes): Phase-2 fd-dir `EACCES → 'owned'` over-skips on multi-user / root hosts
`/proc/<pid>/comm` and `/cmdline` are world-readable (0444) but `/proc/<pid>/fd` is owner-only (0500, ptrace-gated). So a **cross-user or root-run `gitnexus mcp`/`serve` process serving *any* repo** passes Phase 0 + Phase 1, hits `EACCES` at Phase 2, and returns `'owned'` → this repo's augment is silently skipped, though that server never holds this repo's lbug. The old code did `catch { continue; }` *and* ran the cmdline check **after** the fd-holds check, so it did **not** over-skip this way. This is fail-closed (the safe direction for DB-lock avoidance) and the inline comment *does* acknowledge the cross-user case — three things still make it worth a second look:
- (a) the candidate was **never shown to hold *this* lbug** (the dev+inode check is skipped on this path), so it's any foreign gitnexus process, not one proven relevant;
- (b) the `catch` returns `'owned'` for **every** non-`ENOENT` error (`ENOTDIR`/`EIO`/`EPERM`/`ESTALE`), some transient and better treated as race / not-owner;
- (c) **no operator diagnostic** — unlike other skip paths this one emits nothing, and the adapter's existing `GITNEXUS_DEBUG` line says "MCP server owns DB" for owned, timeout, and this case alike, so an operator can't tell why augment stopped.

Suggested: return `'timeout'` (not `'owned'`) on `EACCES` — same fail-closed outcome, no false ownership claim — and/or narrow the `catch` to permission errors only, plus a distinct `GITNEXUS_DEBUG` reason.

### ② Inline, P2 (test integrity): the escalation loop ships unexercised, and the test that claims to cover it doesn't
`hook-db-lock-probe.test.ts:201` ("…read escalation, no miss") builds a ~9 KB cmdline (`'/very/long/path'.repeat(600)`) and never sets `GITNEXUS_HOOK_PROC_CMDLINE_MAX`, so under the 16 KB default cap the first `readSync` consumes the whole cmdline in one call and the multi-iteration escalation loop never repeats. **No test in the suite triggers the escalation path.** The escalation *code* is correct (Codex-verified), so this is a coverage/labeling gap, not a live bug — but the path that exists precisely to avoid missing a far-out mode token is unguarded against regression. Fix in-PR: set `GITNEXUS_HOOK_PROC_CMDLINE_MAX: '4096'` with a >4 KB cmdline (`mcp` past the cap) so escalation actually iterates — or correct the test name.

### Lower-priority (body only)
- **Permanent silent degradation**: on a consistently-overloaded host every call times out → augment skipped forever, no circuit-breaker. Documented self-throttle, but an operator signal would help (folds into ①'s diagnostic).
- **`GITNEXUS_HOOK_PROC_ROOT` is read on the production path** though documented test-only — a stray/leaked env export silently disables Linux owner detection (fail-**open** → augment contends with a real server). Consider gating behind a test signal.
- **Coverage shift**: Linux owner-detection is now in-process unit-tested (fake `/proc`) only; the spawned-hook owner-skip e2e now runs on macOS (lsof lane). No Linux end-to-end asserts the real dispatcher skips augment on a detected owner.
- **Escalation cost** (adversarial flagged a DoS; perf bounded it): per-candidate cost is `HARD_CEIL`-bounded (256 KiB, ~16 chunks, ~2.5 ms). Note `outOfBudget()` is **not** re-checked inside the `readLinuxCmdline` escalation loop, so one pathological candidate runs its full escalation before the next budget check — still bounded (overrun ≈ one candidate's escalation), hence **P3 hardening** (add an in-loop budget check), not a real DoS.
- **Live `/proc` e2e flake risk**: the 5 s holder-startup poll on a shared CI runner could intermittently false-negative.
- Nits: budget env `parseInt` ignores scientific notation (`16e3`→16 ms); `Buffer.alloc`→`allocUnsafe` on the hot read; defensive `fd >= 0` guard before `closeSync`.

### CI
Green — the only red, **Vercel**, is "authorization required", not a code result.

---
_Automated multi-tool digest (GitNexus swarm + Compound-Engineering personas + Codex), gated through an independent synthesis critic. Verify before acting. The headline is the `EACCES` fail-closed semantics; everything else validated cleanly._

File: gitnexus/hooks/claude/hook-db-lock-probe.cjs (hunk 1 above)
Comment (Minidoracat) [reply gitnexus/hooks/claude/hook-db-lock-probe.cjs]: Fixed in d5366cd8 — the catch no longer claims `'owned'` for an unverifiable fd dir. Split by errno: `ENOENT` → continue (raced away); `EACCES`/`EPERM` and transient `EIO`/`ESTALE` → `'timeout'` (same fail-closed dispatcher outcome, but honest — no false ownership claim for a server whose dev+ino was never compared against this lbug); `ENOTDIR`/other structural errors → continue (not a real fd dir). Added a `GITNEXUS_DEBUG` line on the EACCES path so an operator can distinguish this inconclusive skip from a genuine owner.

On fail-closed vs fail-open here: I kept it fail-closed per your reasoning and Codex's — augment is optional context, so over-skipping costs a little enrichment, whereas fail-open against a real cross-user owner re-opens the #1492 lock contention. The change is purely *owned → timeout* (honest), not a flip to fail-open.

New uid-agnostic spy tests (`vi.spyOn(fs,'readdirSync')` throwing each errno) pin all five branches regardless of the runner's uid — the disk chmod-000 tests no-op under root, which is exactly why the EACCES assertion needed a uid-independent path. Mirrored in the byte-identical plugin copy.

File: gitnexus/test/unit/hook-db-lock-probe.test.ts (hunk 2 above)
Comment (Minidoracat) [reply gitnexus/test/unit/hook-db-lock-probe.test.ts]: Fixed in d5366cd8. You (and Codex) were right that just setting the cap wasn't enough — with the long pad *before* the gitnexus token, the first 4 KB chunk wouldn't even see `gitnexus` and the `if (!hasGitNexus) break` guard would stop the loop before it escalated. Reshaped the cmdline so the gitnexus token sits under 4 KB and the `mcp` mode token is padded past `GITNEXUS_HOOK_PROC_CMDLINE_MAX=4096`, and added a `readSync` spy asserting >1 read so the escalation is proven to iterate. While there, the escalation loop now re-checks the budget each iteration and returns a distinct timeout sentinel — never `''` (an empty string reads as 'not a candidate' and could drop a real owner → fail-open); the caller maps the sentinel to `'timeout'`.

**Comment by Minidoracat** [comment]
## Review follow-up — both P2s + lower-priority addressed in d5366cd8

Thanks for the careful pass (and for running it through Codex — its two refinements were load-bearing). Dispositions:

**Inline P2s** — both fixed, see inline replies:
1. **EACCES → 'owned' over-skip** — split by errno; EACCES/EPERM/EIO/ESTALE → `'timeout'` (honest fail-closed, no false ownership), ENOTDIR/ENOENT/other → continue; `GITNEXUS_DEBUG` diagnostic; uid-agnostic spy tests pin every branch.
2. **Escalation test didn't escalate** — reshaped so gitnexus is in the first 4 KB chunk and `mcp` is past the 4096 cap, with a `readSync` spy proving >1 read.

**Lower-priority:**
- **(a) escalation loop budget check** — done. The loop re-checks the budget each iteration and returns a distinct timeout sentinel (never `''`, which would read as "not a candidate" and could drop a real owner → fail-open); caller maps it to `'timeout'`.
- **(b) `GITNEXUS_HOOK_PROC_ROOT` on the production path** — gated to test context, so a stray production env export can no longer disable Linux owner detection; the fake-procfs tests still inject it.
- **(c) `parseInt` ignores scientific notation (`16e3`→16)** — left as a separate follow-up; it's a pre-existing parse footgun on an integer-ms env, orthogonal to this PR's owner-detection logic. Happy to fold it in if you'd prefer.

**Note on the EIO/ESTALE → timeout split:** the original suggestion lumped all non-ENOENT-non-EACCES errors as "continue", but I treat transient I/O (EIO/ESTALE) as `'timeout'` rather than continue — a mid-I/O-blip on a real owner's fd dir shouldn't be read as "this candidate isn't the owner" (that would be a fail-open miss). ENOTDIR stays continue (structurally not an fd dir). Open to collapsing them if you'd rather keep it minimal.

Suite after the change: probe + hooks 177 passed / 36 skipped (the relaned lsof lanes). Rebased onto your `main` merge (ee54c1dc), probe copies byte-identical.

File: gitnexus/hooks/claude/hook-db-lock-probe.cjs [hunk 3]
```diff
@@ -242,59 +274,313 @@
...
+      const hasMode = /(?:^|\s)(mcp|serve)(?:\s|$)/.test(text);
+      if (hasMode) break; // decided (positive); isGitNexusServerCommand re-checks below
+      if (bytes < chunkCap) break; // EOF: full cmdline read, definitive
+      if (!hasGitNexus) break; // not a candidate; do not escalate the read
+      if (offset >= HARD_CEIL) break; // bounded escalation only
+      // Budget gate the escalation: a single huge-argv candidate must not burn
+      // the whole scan deadline before we re-check. Return the timeout sentinel
+      // (never '') so the caller fails closed instead of treating us as a
+      // non-candidate.
+      if (typeof outOfBudget === 'function' && outOfBudget()) return CMDLINE_TIMEOUT;
```
[the remark this answers is not in the document]
Comment (Minidoracat) [reply gitnexus/hooks/claude/hook-db-lock-probe.cjs]: Fixed in 5dfe268d. Correct — `readLinuxCmdline` has a single caller (`linuxProcScanFindGitNexusServer`) that always passes `outOfBudget`, so the `typeof … === 'function'` guard was dead. Dropped it (now `if (outOfBudget())`) and noted the invariant. Mirrored in the byte-identical plugin copy.
